File: tools/modplatform/bindings.py
#!/usr/bin/env python3
"""The binding profile: everything about ONE game build that cannot be resolved.

WHAT BELONGS IN HERE, AND WHY THE LINE IS WHERE IT IS
-----------------------------------------------------
Stage 5B splits what the runtime needs to know in two:

  * **build-specific measured facts** -> this profile. Addresses that exist only
    because this executable was linked this way, vtable slot indices, the
    ParmsSize a reflected function must have, the offset a struct field must sit
    at. None of these can be discovered from a running process without guessing,
    and guessing is exactly what an unsupported build must never get.
  * **per-run dynamic facts** -> ``runtime/MiseryRuntime/Internal/Resolver``.
    Object pointers, UFunction addresses, live instances. These differ every
    launch, so a profile could not carry them even in principle.

The two meet at the fail-closed check: the resolver finds something, the profile
says what it must be, and a disagreement stops the runtime instead of being
absorbed. A profile is therefore not a convenience -- it is the thing that turns
"we found an object called ItemList" into "we found the ItemList this framework
was built against".

WHY THE PROFILE IS EMITTED RATHER THAN HAND-WRITTEN
---------------------------------------------------
Every number in here already exists somewhere in this repository, attached to
the run that measured it. Re-typing them into a JSON file would create a second
copy that nothing compares, which is how the ModId rule drifted across three
stages. So the emitter READS the measured constants and records, per fact, where
it read them from; ``tests/test_bindings.py`` then asserts the emitted profile
still equals its sources. Drift becomes a failing test rather than a wrong
address in a shipping game.

WHY THIS TOOL MAY IMPORT FROM research/
---------------------------------------
It is a build-time emitter, not part of the runtime. What ships is the JSON it
produces; nothing under research/ is loaded by the game. The dependency runs in
the safe direction: the tool that writes the profile reads the record that
proved it.

CODE BYTES ARE THE ACTUAL GUARD
-------------------------------
An RVA on its own is a number that used to be right. Each code address in the
profile therefore carries the first 16 bytes found at it, read from the shipped
executable. The runtime compares those bytes against live memory before it uses
the address for anything. A Steam patch that moves a function does not produce a
subtly wrong call -- it produces a byte mismatch and a refusal.
"""
import argparse
import datetime
import hashlib
import json
import os
import struct
import sys

TOOL_VERSION = "1.0.0"
BINDINGS_VERSION = 1

REPO = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
for _p in (os.path.join(REPO, "research", "instruments", "eri"),
           os.path.join(REPO, "research", "instruments", "ipp")):
    if _p not in sys.path:
        sys.path.insert(0, _p)


# ---------------------------------------------------------------- sources ----
def measured_sources():
    """Every build-specific constant, with the module that measured it.

    Imported rather than copied. If a research constant moves, this emitter
    moves with it and the test that compares them stays green; if somebody edits
    the emitted JSON by hand instead, that same test goes red.
    """
    import eri                                                  # noqa: E402
    import cr01c1_controller as c1                              # noqa: E402
    import fts_controller as fts                                # noqa: E402

    # Code addresses: verified byte-for-byte against live memory before use.
    code = {
        "add_ticker": (fts.RVA_ADD_TICKER, "research/instruments/ipp/fts_controller.py"),
        "get_core_ticker": (fts.RVA_GET_CORE_TICKER,
                            "research/instruments/ipp/fts_controller.py"),
        "fmemory_malloc": (fts.RVA_FMEMORY_MALLOC,
                           "research/instruments/ipp/fts_controller.py"),
        # Measured in CR-01C5's resolve(): base + these, all eight byte-verified
        # live == disk before the probe would run.
        "fmemory_free": (0xFA0090, "research/evidence/CR-01C5 resolve()"),
        "set_root_flags": (0x1210E60, "research/evidence/CR-01C5 resolve()"),
        "clear_root_flags": (0x11BB340, "research/evidence/CR-01C5 resolve()"),
    }
    # Data addresses: there are no stable bytes at a live global, so these are
    # checked structurally by the resolver instead (a chunk count that is not
    # plausible, or a name block that will not decode, fails the same way).
    data = {
        "guobjectarray": (eri.DEFAULT_GUOBJECTARRAY_RVA,
                          "research/instruments/eri/eri.py"),
        "namepool": (eri.DEFAULT_NAMEPOOL_RVA, "research/instruments/eri/eri.py"),
        "name_pool_initialized": (eri.DEFAULT_NAME_POOL_INITIALIZED_RVA,
                                  "research/instruments/eri/eri.py"),
    }
    slots = {
        # Proven by agreement across three unrelated CDOs rather than by one
        # lucky read -- see CR-01C1.
        "process_event": (c1.PE_SLOT, "research/instruments/ipp/cr01c1_controller.py"),
        "datatable_add_row": (95, "research/evidence/CR-01C4B"),
        "datatable_remove_row": (94, "research/evidence/CR-01C4B"),
        "scriptstruct_initialize": (96, "research/evidence/CR-01C4B"),
        "scriptstruct_destroy": (97, "research/evidence/CR-01C4B"),
    }
    return code, data, slots


# Reflected functions, and what each must look like before it may be called.
# Copied from the gate() calls CR-01C5 makes: the same numbers, in the file that
# now owns them for production, and compared against that controller by a test.
#
# FORBID is the net/authority mask. A function carrying any of those bits is one
# the engine may replicate, and calling it locally would be a gameplay decision
# made by accident.
FUNCTION_FORBID_FLAGS = 0x0138C0C4
FUNCTION_GATES = {
    "GameplayStatics::SpawnObject":
        {"parms_size": 24, "return_value_offset": None, "require_flags": 0x2400},
    "KismetTextLibrary::Conv_StringToText":
        {"parms_size": 32, "return_value_offset": 16, "require_flags": 0x2400},
    "KismetTextLibrary::Conv_TextToString":
        {"parms_size": 32, "return_value_offset": 16, "require_flags": 0x2400},
    "KismetSystemLibrary::LoadAsset_Blocking":
        {"parms_size": 48, "return_value_offset": 40, "require_flags": 0x2400},
    "KismetSystemLibrary::Conv_SoftObjectReferenceToString":
        {"parms_size": 56, "return_value_offset": 40, "require_flags": 0x2400},
    "BP_SGKFunctions_C::SGK ItemDetails":
        {"parms_size": 2336, "return_value_offset": None, "require_flags": 0},
    "BP_MasterInventory_C::AddItem":
        {"parms_size": 120, "return_value_offset": None, "require_flags": 0},
    "BP_MasterInventory_C::RemoveItem":
        {"parms_size": 83, "return_value_offset": None, "require_flags": 0},
}

# The row struct the Items backend writes. Offsets are read from the CR-01B
# reflection dump rather than restated, and only the fields the framework
# actually touches are carried: a profile that listed all 32 would be asserting
# things no code checks.
ROW_STRUCT_NAME = "S_ItemDetails"
ROW_STRUCT_FIELDS = ("Name", "ShortName", "Description", "Weight", "Width",
                     "Height", "AllowStacking", "MaxStack", "StaticMesh",
                     "WorldClass", "UIDetails", "ItemOffsets")
STRUCT_DEFS = os.path.join(REPO, "research", "evidence", "CR-01B",
                           "structs-defs.json")
# The struct's PropertiesSize, from the run that wrote a row through engine
# AddRow and hashed all 496 vanilla rows at this width. The resolver reads the
# live value; this is what it must equal.
ROW_STRUCT_SIZE_SOURCE = os.path.join(REPO, "research", "evidence", "CR-01C5",
                                      "retired-demo-state.json")

# UDataTable member offsets. UE 5.4 layout facts, restated here because the
# runtime reads them through the profile and a version mismatch must be visible
# in one place.
OBJECT_LAYOUT = {
    "datatable_rowstruct": 40,
    "datatable_parent_tables": 176,
    "ustruct_properties_size": 0x58,
}


# ------------------------------------------------------------------- PE ------
def section_bytes(image, rva, count):
    """The bytes at *rva*, found through the PE section table.

    The same walk the research controllers use. It returns None rather than
    guessing when an RVA falls outside every section's raw data, because an
    address that is not backed by file data is not an address this profile can
    make a promise about.
    """
    pe = struct.unpack_from("<I", image, 0x3C)[0]
    sections = struct.unpack_from("<H", image, pe + 6)[0]
    opt_size = struct.unpack_from("<H", image, pe + 20)[0]
    table = pe + 24 + opt_size
    for index in range(sections):
        entry = table + index * 40
        virt_size, virt_addr, raw_size, raw_ptr = struct.unpack_from(
            "<IIII", image, entry + 8)
        if virt_addr <= rva < virt_addr + max(virt_size, raw_size) and \
                rva - virt_addr < raw_size:
            start = raw_ptr + (rva - virt_addr)
            return image[start:start + count]
    return None


def image_size(image):
    pe = struct.unpack_from("<I", image, 0x3C)[0]
    return struct.unpack_from("<I", image, pe + 24 + 56)[0]


# -------------------------------------------------------------- emitting ----
class BindingsError(Exception):
    pass


def row_struct_size():
    with open(ROW_STRUCT_SIZE_SOURCE, encoding="utf-8") as handle:
        state = json.load(handle)
    size = state.get("struct_size")
    if not isinstance(size, int) or size <= 0:
        raise BindingsError("%s states no usable struct_size"
                            % ROW_STRUCT_SIZE_SOURCE)
    return size


def row_struct_offsets():
    with open(STRUCT_DEFS, encoding="utf-8") as handle:
        defs = json.load(handle)
    if ROW_STRUCT_NAME not in defs:
        raise BindingsError("%s is not in %s" % (ROW_STRUCT_NAME, STRUCT_DEFS))
    # Field names carry a GUID suffix the cooker generates; the logical name is
    # the part before the first underscore, which is what the resolver matches
    # on and what the profile therefore records.
    found = {}
    for field in defs[ROW_STRUCT_NAME]["fields"]:
        logical = field["name"].split("_")[0]
        if logical in found:
            raise BindingsError(
                "two fields of %s share the logical name %r -- the prefix rule "
                "does not identify them and the profile will not guess"
                % (ROW_STRUCT_NAME, logical))
        found[logical] = int(field["offset"])
    missing = [name for name in ROW_STRUCT_FIELDS if name not in found]
    if missing:
        raise BindingsError("%s has no field(s) %s" % (ROW_STRUCT_NAME, missing))
    return {name: found[name] for name in ROW_STRUCT_FIELDS}


def emit(exe_path, build_id, build_key, engine, *, generated_at=None):
    """Build the profile for the executable at *exe_path*.

    Nothing here trusts the caller's *build_key*: the digest is recomputed from
    the file, and a mismatch is refused. A profile whose key came from anywhere
    but the bytes it describes is a profile that can be pointed at the wrong
    build.
    """
    with open(exe_path, "rb") as handle:
        image = handle.read()
    digest = "sha256:" + hashlib.sha256(image).hexdigest()
    if build_key and digest != build_key:
        raise BindingsError("%s hashes to %s, not the requested %s"
                            % (exe_path, digest, build_key))

    code, data, slots = measured_sources()
    addresses = {}
    for name, (rva, source) in sorted(code.items()):
        raw = section_bytes(image, rva, 16)
        if raw is None:
            raise BindingsError("rva 0x%X (%s) is not backed by file data"
                                % (rva, name))
        addresses[name] = {"kind": "code", "rva": rva,
                           "bytes": raw.hex(), "source": source}
    for name, (rva, source) in sorted(data.items()):
        addresses[name] = {"kind": "data", "rva": rva, "source": source}

    profile = {
        "bindings_version": BINDINGS_VERSION,
        "generated_by": "tools/modplatform/bindings.py %s" % TOOL_VERSION,
        "generated_at": generated_at or (
            datetime.datetime.now(datetime.timezone.utc)
            .strftime("%Y-%m-%dT%H:%M:%SZ")),
        "build": {
            "build_id": build_id,
            "build_key": digest,
            "image_size_bytes": image_size(image),
            "file_size_bytes": len(image),
            "engine_version": engine["engine_version"],
            "engine_cl": engine["engine_cl"],
            "build_configuration": engine["build_configuration"],
        },
        "addresses": addresses,
        "vtable_slots": {name: {"slot": value, "source": source}
                         for name, (value, source) in sorted(slots.items())},
        "functions": {
            "forbid_flags": FUNCTION_FORBID_FLAGS,
            "gates": {name: dict(gate)
                      for name, gate in sorted(FUNCTION_GATES.items())},
        },
        "row_struct": {
            "name": ROW_STRUCT_NAME,
            "size": row_struct_size(),
            "fields": row_struct_offsets(),
            "source": "research/evidence/CR-01B/structs-defs.json",
            "size_source": "research/evidence/CR-01C5/retired-demo-state.json",
        },
        "object_layout": dict(OBJECT_LAYOUT),
    }
    return profile


def engine_from_index(build_key):
    """Engine identity, read from the build registry rather than restated."""
    path = os.path.join(REPO, "research", "builds", "index.json")
    with open(path, encoding="utf-8") as handle:
        index = json.load(handle)
    if build_key not in index:
        raise BindingsError("%s is not a registered build" % build_key)
    entry = index[build_key]
    fingerprint_rel = entry.get("artifacts", {}).get("fingerprint_json")
    engine = {"engine_version": entry.get("engine_version"),
              "engine_cl": None, "build_configuration": None}
    if fingerprint_rel:
        with open(os.path.join(REPO, fingerprint_rel), encoding="utf-8") as handle:
            fingerprint = json.load(handle)
        group = fingerprint.get("engine") or {}
        engine["engine_cl"] = group.get("engine_cl")
        engine["build_configuration"] = group.get("build_configuration")
    for key, value in engine.items():
        if value in (None, ""):
            raise BindingsError("the build registry does not state %s for %s"
                                % (key, build_key))
    return entry["build_id"], engine


def write(profile, out_path):
    os.makedirs(os.path.dirname(os.path.abspath(out_path)), exist_ok=True)
    with open(out_path, "w", encoding="utf-8", newline="\n") as handle:
        json.dump(profile, handle, indent=2, sort_keys=True)
        handle.write("\n")
    return out_path


def main(argv=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--exe", required=True,
                        help="MISERY-Win64-Shipping.exe to describe")
    parser.add_argument("--out", required=True)
    parser.add_argument("--build-key", default=None,
                        help="expected sha256:...; recomputed and checked")
    args = parser.parse_args(argv)

    with open(args.exe, "rb") as handle:
        digest = "sha256:" + hashlib.sha256(handle.read()).hexdigest()
    build_key = args.build_key or digest
    build_id, engine = engine_from_index(build_key)
    profile = emit(args.exe, build_id, build_key, engine)
    write(profile, args.out)
    print(json.dumps({"out": args.out, "build_id": build_id,
                      "build_key": profile["build"]["build_key"],
                      "addresses": len(profile["addresses"]),
                      "gates": len(profile["functions"]["gates"])}, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: runtime/MiseryRuntime/Internal/Bindings.cpp
#include "Bindings.h"

#include <stdarg.h>
#include <stdio.h>
#include <string.h>
#include <windows.h>

#include "Json.h"
#include "Resolver.h"

namespace misery {
namespace bindings {

const char kSupportedEngineVersion[] = "5.4.4";

namespace {

// A profile is a few kilobytes. Anything larger is not the file we wrote.
constexpr size_t kMaxProfileBytes = 1u << 20;

void Say(std::string* error, const char* format, ...) {
  if (error == nullptr || !error->empty()) {
    return;
  }
  char buffer[512];
  va_list args;
  va_start(args, format);
  _vsnprintf_s(buffer, sizeof(buffer), _TRUNCATE, format, args);
  va_end(args);
  *error = buffer;
}

const json::Value* Need(const json::Value& parent, const char* key,
                        json::Kind kind, const char* where,
                        std::string* error) {
  const json::Value* member = parent.Member(key);
  if (member == nullptr) {
    Say(error, "the profile has no %s.%s", where, key);
    return nullptr;
  }
  if (!member->Is(kind)) {
    // THE PROFILE CARRIES ONLY INTEGERS, and this is where that is enforced.
    //
    // It used to be enforced one layer down: the JSON parser refused any
    // fractional number outright. The parser now reads floats, because a mod's
    // settings file has a float type (Stage 8.5) and one parser serves both
    // documents. The promise about the PROFILE did not change, so it moved to
    // the reader that made it -- and it names the reason, because "not of the
    // expected type" would hide that a number was written where an integer
    // belongs.
    if (kind == json::Kind::kInt && member->Is(json::Kind::kDouble)) {
      Say(error, "%s.%s is a fractional number; the profile carries only "
                 "integers", where, key);
      return nullptr;
    }
    Say(error, "%s.%s is not of the expected type", where, key);
    return nullptr;
  }
  return member;
}

bool NeedString(const json::Value& parent, const char* key, const char* where,
                std::string* out, std::string* error) {
  const json::Value* member = Need(parent, key, json::Kind::kString, where,
                                   error);
  if (member == nullptr) {
    return false;
  }
  *out = member->text;
  return true;
}

bool NeedU32(const json::Value& parent, const char* key, const char* where,
             uint32_t* out, std::string* error) {
  const json::Value* member = Need(parent, key, json::Kind::kInt, where, error);
  if (member == nullptr) {
    return false;
  }
  if (member->integer < 0 || member->integer > 0xFFFFFFFFll) {
    Say(error, "%s.%s does not fit in 32 bits", where, key);
    return false;
  }
  *out = static_cast<uint32_t>(member->integer);
  return true;
}

// "48895c..." -> 16 bytes. Exactly 32 hex digits; a short or odd-length string
// is a malformed profile, not something to pad.
//
// *which* is threaded through so the message names the address. Say() keeps the
// FIRST error, so a generic message here would be the one the user sees and a
// later "addresses.<name> is unreadable" would never land.
bool ParseHex16(const std::string& hex, const char* which, uint8_t* out,
                std::string* error) {
  if (hex.size() != 32) {
    Say(error, "addresses.%s: expected 32 hex characters of code bytes, got %zu",
        which, hex.size());
    return false;
  }
  for (size_t i = 0; i < 16; ++i) {
    uint32_t value = 0;
    for (size_t half = 0; half < 2; ++half) {
      char c = hex[i * 2 + half];
      uint32_t digit;
      if (c >= '0' && c <= '9') {
        digit = static_cast<uint32_t>(c - '0');
      } else if (c >= 'a' && c <= 'f') {
        digit = static_cast<uint32_t>(c - 'a' + 10);
      } else if (c >= 'A' && c <= 'F') {
        digit = static_cast<uint32_t>(c - 'A' + 10);
      } else {
        Say(error, "addresses.%s: the expected code bytes contain a non-hex "
                   "character", which);
        return false;
      }
      value = (value << 4) | digit;
    }
    out[i] = static_cast<uint8_t>(value);
  }
  return true;
}

}  // namespace

bool Profile::Rva(const std::string& name, uint64_t* out) const {
  auto it = addresses.find(name);
  if (it == addresses.end()) {
    return false;
  }
  *out = it->second.rva;
  return true;
}

bool Profile::Slot(const std::string& name, uint32_t* out) const {
  auto it = vtable_slots.find(name);
  if (it == vtable_slots.end()) {
    return false;
  }
  *out = it->second;
  return true;
}

bool Profile::Field(const std::string& name, uint32_t* out) const {
  auto it = row_struct_fields.find(name);
  if (it == row_struct_fields.end()) {
    return false;
  }
  *out = it->second;
  return true;
}

bool Profile::WriteOffset(const std::string& name, uint32_t* out) const {
  auto it = write_offsets.find(name);
  if (it == write_offsets.end()) {
    return false;
  }
  *out = it->second;
  return true;
}

bool Profile::InventoryOffset(const std::string& name, uint32_t* out) const {
  auto it = inventory.find(name);
  if (it == inventory.end()) {
    return false;
  }
  *out = it->second;
  return true;
}

const FunctionGate* Profile::Gate(const std::string& qualified_name) const {
  auto it = function_gates.find(qualified_name);
  return it == function_gates.end() ? nullptr : &it->second;
}

bool Load(const char* path, const char* expected_build_key, Profile* out,
          std::string* error) {
  std::string sink;
  if (error == nullptr) {
    error = &sink;
  }
  error->clear();
  std::string text;
  if (!json::ReadFile(path, kMaxProfileBytes, &text, error)) {
    return false;
  }
  json::Value root;
  std::string parse_error;
  if (!json::Parse(text, &root, &parse_error)) {
    Say(error, "the profile is not readable JSON: %s", parse_error.c_str());
    return false;
  }
  if (!root.Is(json::Kind::kObject)) {
    Say(error, "the profile is not a JSON object");
    return false;
  }

  const json::Value* version = Need(root, "bindings_version", json::Kind::kInt,
                                    "the profile", error);
  if (version == nullptr) {
    return false;
  }
  if (version->integer != kSupportedBindingsVersion) {
    Say(error, "the profile states bindings_version %lld; this runtime "
               "understands %lld and will not read another version leniently",
        static_cast<long long>(version->integer),
        static_cast<long long>(kSupportedBindingsVersion));
    return false;
  }

  // ---- identity, checked against the digest the bootstrap measured -------
  const json::Value* build = Need(root, "build", json::Kind::kObject,
                                  "the profile", error);
  if (build == nullptr) {
    return false;
  }
  if (!NeedString(*build, "build_key", "build", &out->build_key, error) ||
      !NeedString(*build, "build_id", "build", &out->build_id, error) ||
      !NeedString(*build, "engine_version", "build", &out->engine_version,
                  error) ||
      !NeedString(*build, "build_configuration", "build",
                  &out->build_configuration, error)) {
    return false;
  }
  const json::Value* cl = Need(*build, "engine_cl", json::Kind::kInt, "build",
                               error);
  if (cl == nullptr) {
    return false;
  }
  out->engine_cl = cl->integer;
  const json::Value* image = Need(*build, "image_size_bytes", json::Kind::kInt,
                                  "build", error);
  if (image == nullptr || image->integer <= 0) {
    Say(error, "build.image_size_bytes is not a positive integer");
    return false;
  }
  out->image_size_bytes = static_cast<uint64_t>(image->integer);

  if (expected_build_key != nullptr && *expected_build_key != '\0' &&
      out->build_key != expected_build_key) {
    Say(error, "the profile describes %s, but this executable hashes to %s",
        out->build_key.c_str(), expected_build_key);
    return false;
  }
  if (out->engine_version != kSupportedEngineVersion) {
    Say(error, "the profile is for engine %s; this runtime's compiled-in type "
               "layout describes %s and the profile cannot correct it",
        out->engine_version.c_str(), kSupportedEngineVersion);
    return false;
  }

  // ---- addresses ---------------------------------------------------------
  const json::Value* addresses = Need(root, "addresses", json::Kind::kObject,
                                      "the profile", error);
  if (addresses == nullptr) {
    return false;
  }
  for (const auto& entry : addresses->object) {
    const json::Value& value = entry.second;
    if (!value.Is(json::Kind::kObject)) {
      Say(error, "addresses.%s is not an object", entry.first.c_str());
      return false;
    }
    std::string kind;
    if (!NeedString(value, "kind", entry.first.c_str(), &kind, error)) {
      return false;
    }
    const json::Value* rva = Need(value, "rva", json::Kind::kInt,
                                  entry.first.c_str(), error);
    if (rva == nullptr) {
      return false;
    }
    if (rva->integer < 0 || static_cast<uint64_t>(rva->integer) >=
                                out->image_size_bytes) {
      Say(error, "addresses.%s has an rva outside the image",
          entry.first.c_str());
      return false;
    }
    Address address;
    address.rva = static_cast<uint64_t>(rva->integer);
    if (kind == "code") {
      address.is_code = true;
      std::string hex;
      if (!NeedString(value, "bytes", entry.first.c_str(), &hex, error)) {
        return false;
      }
      if (!ParseHex16(hex, entry.first.c_str(), address.expected, error)) {
        return false;
      }
      address.has_expected = true;
    } else if (kind != "data") {
      Say(error, "addresses.%s has kind '%s', which is neither code nor data",
          entry.first.c_str(), kind.c_str());
      return false;
    }
    out->addresses[entry.first] = address;
  }

  // ---- vtable slots ------------------------------------------------------
  const json::Value* slots = Need(root, "vtable_slots", json::Kind::kObject,
                                  "the profile", error);
  if (slots == nullptr) {
    return false;
  }
  for (const auto& entry : slots->object) {
    if (!entry.second.Is(json::Kind::kObject)) {
      Say(error, "vtable_slots.%s is not an object", entry.first.c_str());
      return false;
    }
    uint32_t slot = 0;
    if (!NeedU32(entry.second, "slot", entry.first.c_str(), &slot, error)) {
      return false;
    }
    // A vtable index far past any plausible UObject vtable is a typo, and a
    // typo here reads an arbitrary pointer and calls it.
    if (slot > 512) {
      Say(error, "vtable_slots.%s is slot %u, past any plausible vtable",
          entry.first.c_str(), slot);
      return false;
    }
    out->vtable_slots[entry.first] = slot;
  }

  // ---- function gates ----------------------------------------------------
  const json::Value* functions = Need(root, "functions", json::Kind::kObject,
                                      "the profile", error);
  if (functions == nullptr) {
    return false;
  }
  if (!NeedU32(*functions, "forbid_flags", "functions",
               &out->function_forbid_flags, error)) {
    return false;
  }
  const json::Value* gates = Need(*functions, "gates", json::Kind::kObject,
                                  "functions", error);
  if (gates == nullptr) {
    return false;
  }
  for (const auto& entry : gates->object) {
    if (!entry.second.Is(json::Kind::kObject)) {
      Say(error, "functions.gates.%s is not an object", entry.first.c_str());
      return false;
    }
    FunctionGate gate;
    if (!NeedU32(entry.second, "parms_size", entry.first.c_str(),
                 &gate.parms_size, error) ||
        !NeedU32(entry.second, "require_flags", entry.first.c_str(),
                 &gate.require_flags, error)) {
      return false;
    }
    const json::Value* rvo = entry.second.Member("return_value_offset");
    if (rvo == nullptr) {
      Say(error, "functions.gates.%s states no return_value_offset (use null "
                 "to mean unconstrained)", entry.first.c_str());
      return false;
    }
    if (rvo->Is(json::Kind::kInt)) {
      gate.return_value_offset = rvo->integer;
    } else if (!rvo->Is(json::Kind::kNull)) {
      Say(error, "functions.gates.%s has a return_value_offset that is neither "
                 "an integer nor null", entry.first.c_str());
      return false;
    }
    out->function_gates[entry.first] = gate;
  }

  // ---- the row struct ----------------------------------------------------
  const json::Value* row = Need(root, "row_struct", json::Kind::kObject,
                                "the profile", error);
  if (row == nullptr) {
    return false;
  }
  if (!NeedString(*row, "name", "row_struct", &out->row_struct_name, error) ||
      !NeedU32(*row, "size", "row_struct", &out->row_struct_size, error)) {
    return false;
  }
  if (out->row_struct_size == 0) {
    Say(error, "row_struct.size is zero");
    return false;
  }
  const json::Value* fields = Need(*row, "fields", json::Kind::kObject,
                                   "row_struct", error);
  if (fields == nullptr) {
    return false;
  }
  for (const auto& entry : fields->object) {
    if (!entry.second.Is(json::Kind::kInt) || entry.second.integer < 0 ||
        static_cast<uint64_t>(entry.second.integer) >= out->row_struct_size) {
      Say(error, "row_struct.fields.%s is not an offset inside the struct",
          entry.first.c_str());
      return false;
    }
    out->row_struct_fields[entry.first] =
        static_cast<uint32_t>(entry.second.integer);
  }
  if (out->row_struct_fields.empty()) {
    Say(error, "row_struct.fields is empty");
    return false;
  }

  // The write offsets, and the inventory members beside them. Required, not
  // optional: a runtime that loaded a profile without them would come up and
  // then refuse every registration, which is a worse failure than not starting.
  const json::Value* written = Need(*row, "write_offsets", json::Kind::kObject,
                                    "row_struct", error);
  if (written == nullptr) {
    return false;
  }
  for (const auto& entry : written->object) {
    if (!entry.second.Is(json::Kind::kInt) || entry.second.integer < 0 ||
        static_cast<uint64_t>(entry.second.integer) >= out->row_struct_size) {
      Say(error, "row_struct.write_offsets.%s is not an offset inside the "
                 "struct", entry.first.c_str());
      return false;
    }
    out->write_offsets[entry.first] =
        static_cast<uint32_t>(entry.second.integer);
  }
  const json::Value* inventory = Need(root, "inventory", json::Kind::kObject,
                                      "the profile", error);
  if (inventory == nullptr) {
    return false;
  }
  for (const auto& entry : inventory->object) {
    if (!entry.second.Is(json::Kind::kInt) || entry.second.integer < 0 ||
        entry.second.integer > 0xFFFF) {
      Say(error, "inventory.%s is not a plausible member offset",
          entry.first.c_str());
      return false;
    }
    out->inventory[entry.first] =
        static_cast<uint32_t>(entry.second.integer);
  }

  // ---- object layout -----------------------------------------------------
  const json::Value* layout = Need(root, "object_layout", json::Kind::kObject,
                                   "the profile", error);
  if (layout == nullptr) {
    return false;
  }
  for (const auto& entry : layout->object) {
    uint32_t value = 0;
    if (!entry.second.Is(json::Kind::kInt) || entry.second.integer < 0 ||
        entry.second.integer > 0xFFFF) {
      Say(error, "object_layout.%s is not a plausible member offset",
          entry.first.c_str());
      return false;
    }
    value = static_cast<uint32_t>(entry.second.integer);
    out->object_layout[entry.first] = value;
  }

  // The compiled-in layout and the profile's copy must agree. They are the same
  // measurement recorded twice, and two records that disagree mean one of them
  // is describing a different build.
  const resolve::Layout compiled;
  struct Pair { const char* key; uint32_t compiled; };
  const Pair pairs[] = {
      {"datatable_rowstruct", compiled.datatable_rowstruct},
      {"datatable_parent_tables", compiled.datatable_parent_tables},
      {"ustruct_properties_size", compiled.ustruct_properties_size},
  };
  for (const Pair& pair : pairs) {
    auto it = out->object_layout.find(pair.key);
    if (it == out->object_layout.end()) {
      Say(error, "object_layout states no %s", pair.key);
      return false;
    }
    if (it->second != pair.compiled) {
      Say(error, "object_layout.%s is %u; this runtime was compiled for %u",
          pair.key, it->second, pair.compiled);
      return false;
    }
  }
  return true;
}

bool VerifyCode(const Profile& profile, uint64_t module_base,
                uint64_t module_size, std::string* error) {
  error->clear();
  if (module_size != profile.image_size_bytes) {
    Say(error, "the loaded image is %llu bytes; the profile describes one of "
               "%llu",
        static_cast<unsigned long long>(module_size),
        static_cast<unsigned long long>(profile.image_size_bytes));
    return false;
  }
  int verified = 0;
  for (const auto& entry : profile.addresses) {
    const Address& address = entry.second;
    if (!address.is_code) {
      continue;
    }
    if (!address.has_expected) {
      Say(error, "addresses.%s is code but carries no expected bytes",
          entry.first.c_str());
      return false;
    }
    uint64_t va = module_base + address.rva;
    if (address.rva + sizeof(address.expected) > module_size) {
      Say(error, "addresses.%s would read past the end of the module",
          entry.first.c_str());
      return false;
    }
    uint8_t live[sizeof(address.expected)] = {};
    if (!resolve::ReadBytes(va, live, sizeof(live))) {
      Say(error, "addresses.%s is not readable at 0x%llx",
          entry.first.c_str(), static_cast<unsigned long long>(va));
      return false;
    }
    if (memcmp(live, address.expected, sizeof(live)) != 0) {
      // Deliberately verbose: this is the message that will be in a user's log
      // the day Steam patches the game, and "bindings mismatch" would not tell
      // them what happened.
      Say(error, "addresses.%s does not hold the code the profile recorded. "
                 "This build is not the one these bindings describe; the "
                 "framework refuses rather than call an address that has moved.",
          entry.first.c_str());
      return false;
    }
    ++verified;
  }
  if (verified == 0) {
    Say(error, "the profile carries no code addresses to verify, so nothing "
               "would be checked before use");
    return false;
  }
  return true;
}

bool Resolve(const Profile& profile, const std::string& name,
             uint64_t module_base, uint64_t module_size, uint64_t* out,
             std::string* error) {
  auto it = profile.addresses.find(name);
  if (it == profile.addresses.end()) {
    Say(error, "the profile states no address called %s", name.c_str());
    return false;
  }
  if (it->second.rva >= module_size) {
    Say(error, "%s resolves outside the module", name.c_str());
    return false;
  }
  *out = module_base + it->second.rva;
  return true;
}

}  // namespace bindings
}  // namespace misery
